Hooks to get the current page's Request and Response objects

vmarta opened the pull request:
To have more of a feature parity with `getProps`, we would like to make the current page's `Request` and `Response` objects to be available via hooks too. This PR simply make them accessible throughout the app by relying on React Context.

So the implementation is straightforward actually (basically I mostly copied what's already in #693). But what I'd like us to focus on is the experience of using the hooks, especially how it differs when they're inside `getProps` vs hooks.

Do the page request and response translate well to React hooks? 🤔 

Ticket: W-11668738

**Todos:**
- [ ] Revert my example code

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- React Context and hooks to share the `Request` and `Response` objects of the current page
- Some example of using the hooks in the template-typescript-minimal 

# How to Test-Drive This PR

```
cd packages/template-typescript-minimal
npm start
```

And then verify in the browser inspector (Network tab):
- the status code of the page response is `201` (now `202`)
- and it has the header `Some-Header: FOOBAR`

![Google Chrome 2022-08-31 at 16 13 45](https://user-images.githubusercontent.com/847300/187800999-8ca14056-0590-4829-884b-5a960ac9506e.png)


# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

vmarta committed:
On server-side, share the page's req, res objects

vmarta committed:
Merge branch 'develop' into page-request-response

vmarta committed:
Tweak naming of hooks

vmarta committed:
An example to explore usage of the hooks

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/browser/main.jsx:
```diff
@@ -73,18 +73,22 @@ export const start = () => {
-                        </AppConfig>
-                    </DeviceContext.Provider>
-                </Router>,
+                <PageRequestContext.Provider value={undefined}>
+                    <PageResponseContext.Provider value={undefined}>
```
Similar to when used within `getProps`, the `Request` and `Response` objects are not available on the client-side. 

However, unlike the hooks version, `getProps` allows you to serialize data, which is then available later on the client-side. This fundamental difference introduces a caveat that properties of Request/Response should not be used for rendering a component. See the example in template-typescript-minimal for more details.

vmarta committed:
Add note

vmarta committed:
Add comment

bendvc commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -0,0 +1,12 @@
+import {useContext} from 'react'
+import {PageRequestContext, PageResponseContext} from '../contexts'
+
+export const usePageRequest = () => useContext(PageRequestContext)
+export const usePageResponse = () => useContext(PageResponseContext)
```
Since this is a public api, I think it's best that you have some jsdoc comments on here to explain the eccentricities of these hooks (e.g. the behaviour on client vs server)

bendvc replied:
If you are keen you can add a sample usage in the comment as well.

bendvc commented on packages/template-typescript-minimal/app/pages/home.tsx:
```diff
@@ -105,6 +113,9 @@ const Home = ({value}: Props) => {
                     <div className="divider"></div>
                 </div>
                 <div className="panel">
+                    {/* NOTE: React would complain that this content rendered server-side does not match the one on client-side */}
+                    {pageReq && <p>Path of the page&apos;s request: {pageReq.path}</p>}
+
```
Good thing you pointed this out. I think it might be worth while to note this caveat in the docs for the hooks. Something along the lines of "_usePageResponse and usePageRequest should not be used for deterministic rendering of you component, this will result in markup mismatches during your applications hydration stage_"

vmarta replied:
If the interface can discourage it, then we won't have to jot down this caveat. I think with the latest update, having a callback for the interface would discourage using the server context within the content of a component.

bendvc commented on packages/pwa-kit-react-sdk/src/ssr/universal/contexts/index.js:
```diff
@@ -1,9 +1,12 @@
-export default React.createContext()
+
+export const DeviceContext = React.createContext()
+export const PageRequestContext = React.createContext()
+export const PageResponseContext = React.createContext()
```
Thoughts on the naming of the contexts and hooks. Calling them "page" request and "page" response to me implies that they are or might be other types of requests/responses. 

Naming is hard.. but do you think that useServerRequest and useServerResponse is a little better? The name almost implies that you probably aren't going to get a value using it on the client.. and having server in the name is a little more connected to what the res/req objects are. 

The following comment isn't completely important, but could we possibly combine the 2 into something called useServerContext.. it might give us more room to grow in the future without wrapping components around components. I'm not married to that tho.

vmarta replied:
I like the name `useServerContext`. Like you said, it's clear that these are coming from the server. And the name is more future-proof, allowing for additional properties if need be.

vmarta committed:
Merge branch 'develop' into page-request-response

vmarta committed:
Rename hooks

So that it's clear that these are coming from the server. And the name is more future-proof, allowing for additional properties if need be.

vmarta committed:
Add jsdoc

vmarta committed:
Passing in a callback to `useServerContext`

vmarta commented on packages/template-typescript-minimal/app/pages/home.tsx:
```diff
@@ -90,6 +91,10 @@ const Home = ({value}: Props) => {
         return () => clearInterval(interval)
     }, [counter, setCounter])
 
+    useServerContext(({res: pageRes}) => {
+        pageRes.status(404)
+    })
```
I'd like to propose this updated interface with a callback. Why callback?
- callers don’t need to check whether on server side or not
- it discourages the server context to be used in the content of a component

bendvc commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -0,0 +1,37 @@
+ * @example
+ * useServerContext(({req, res}) => { res.status(404) })
+ */
+export const useServerContext = (fn) => {
+    const serverContext = useContext(ServerContext)
+    const isOnServer = Boolean(serverContext.req)
+
+    if (isOnServer) {
+        fn(serverContext)
+    }
+}
```
I like how with the change to using a callback we don't have to worry about whether or not the `req` and `res` objects are defined, that alleviates some metal overhead that was previously placed on the developer. 

But there are some things that I don't necessarily like. For the most part _hooks_ return values, with the exception being `useEffect`, so not returning anything with from a hook seems to be an exception and not the rule. I expect it should be thought about with special care if we are to do that. (Not a strong argument) 

The other thing is that accomplishing something like creating a page that displays the serialized response now becomes a little more tricky.... It would have to look a little like this:

```
const Page = () => {
    let myRes
    
    useServerContext(({res}) => {myRes = res})
    
    return <div>{JSON.stringify(myRes)}</div>
}
```

vs 

```
const Page = () => {
    const {res = {}} = useServerContext()
    
    return <div>{JSON.stringify(res)}</div>
}
```

I'm not sure if any one of those dislikes is strong enough reason not to proceed with this approach. But if there are more small things like that, then maybe it will add up to something bigger. 

But most importantly it would be nice to see how this would work with the `useQuery` hook in my branch just to get the feel of how the hook is used, and to see if there are any kind of quirks or anything.

vmarta replied:
About **the hook not returning anything**: yes, it behaves more like `useEffect`, and I want it to be unlike the typical kinds of hooks.

Typically, hooks would return something. And this return value can be used in the rendering of the component html. But in our case, we want to discourage it from being used in the html (to avoid the mismatched html between server and client rendered, as was [commented earlier](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/708#discussion_r961040733)).

That's why I prefer having a callback. It focuses you on the request and response within the callback function, and less on the component html.

vmarta replied:
About **serializing request/response**: currently, with the hook approach, there is no mechanism to serialize the request/response. It's possible to do that with `getProps`, though.

So as a result, if we try to print out the request/response, it would only work on the server side. Then React would complain about the mismatched html between server and client rendered.

vmarta replied:
👀 I'll take a look about integrating the `useServerContext` hook with `useQuery` in the other branch.

bendvc replied:
Both are good points.. thanks for clarifying those. I could get on board with them. 

With regards to usage with `useQuery` the way of setting status codes/headers would be like below? : 

```
const Page = () => {
    const {data, error} = useQuery(['my-query'], async () => api.getProduct(1))
    
    useServerContext((res) => {
        res.statusCode = error ? 404 : 200
    })

    return (
        <Fragment>
            {error ? <ErrorComponent error={error}>} : <Product {...data} />
        </Fragment>
    )
}
```

vmarta replied:
@bendvc yup, that's right.. just like that.

I have made a branch here that demonstrates that it is possible to use both `useServerContext` and `useQuery`: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/713#discussion_r965084561

vmarta commented:
I made a new PR #737
